lib/smart_proxy_remote_execution_ssh/runners/script_runner.rb

```diff
@@ -258,32 +230,38 @@ def indent_multiline(string)
     end
 
     def should_cleanup?
-      @session && !@session.closed? && @cleanup_working_dirs
+      @session && @cleanup_working_dirs
     end
 
-    def session
-      @session ||= begin
-                     @logger.debug("opening session to #{@ssh_user}@#{@host}")
-                     Net::SSH.start(@host, @ssh_user, ssh_options)
-                   end
+    def session(command)
+      @session = true
+
+      in_read, in_write = IO.pipe
+      out_read, out_write = IO.pipe
+      args = []
+      args += [{'SSHPASS' => @ssh_password}, '/usr/bin/sshpass', '-e'] if @ssh_password
+      args += ['/usr/bin/ssh', @host, ssh_options, command].flatten
+      command_pid = spawn(*args, :in => in_read, [:out, :err] => out_write)
+      in_read.close
+      out_write.close
+
+      return in_write, out_read, command_pid
```

Comment:
I'm wondering if you really need two pipes. If one is only used for reads and the other only for writes, can't it use a single pipe?

I also wonder if this use case isn't really what [Open3.popen2e](https://ruby-doc.org/stdlib-2.4.1/libdoc/open3/rdoc/Open3.html#method-c-popen2e) is supposed to cover. Wouldn't that be a much better API to use?

Reply:
> I'm wondering if you really need two pipes.

I think so. As far as I know we cannot make a pipe, pass both ends to the child process, but also keep them and read/write from/to the ends we kept. Or am I missing something?

Reply:
But it immediately closes `in_read` and `out_write`. However, I think popen2e opens the pipes for you so that's probably a lot safer.

Reply:
Can you elaborate on that? It should create a new process in a new thread. You have a `wait_thr` variable (which is an instance of [Process::Status](https://ruby-doc.org/core/Process/Status.html)). You can use `wait_thr.value` and that should block until the process has exited. At least if I read the docs.

Reply:
The problem is that `Process.wait(@command_pid)` is waiting for the process with pid which doesn't exist, because that process is already dead. It throws this error:
```
error while dispatching request to runner 375588db-51c1-4ec5-84b6-e7c0f225aadf:Errno::ECHILD No child processes
```

Reply:
That's a good point, but I think it would be worth supporting a `Process::Status` instance instead of a PID in Dynflow. @adamruzicka?

Reply:
After some out of band communication I think the current implementation can be the right one. However, I do think it would be worth adding a comment that the method is like `Open3.popen2e` but without a separate thread to monitor it. Otherwise some may come in and try to refactor it again.